# Demuxlet tools:

Included in this package are a set of functions to facilitate Mux-seq applications. For more information on Demuxlet and Mux-sequencing, see the [Demuxlet GitHub Page](https://github.com/statgen/demuxlet). (Impetus: Many Mux-seq experiments will involve generating the side-by-side bulk and single-cell RNAseq data like the rest of the package is built for.)

## Seurat Import Functions:

These functions generate Seurat objects with Demuxlet information added to metadata.  If you already have you data combined into a single Seurat, they will simply add metadata.  A short summary of SNP counts and doublet calls is also printed at the end.

-	**`Import10XDemux()`** - Imports from 10X cellranger outputs directly. Then Just point the function to the location(s) where your matrix/barcodes/genes files are, how you would like each separate CellRanger input to be named, and the location of your Demuxlet .best output.

-	**`ImportSeuratDemux`** - For importing demuxlet information into datasets generated by any fashion other than 10X, you just need to get the data into Seurat form beforehand; one Seurat per lane/pool/etc.  Point the function to the proper Seurat objects, how you would like each separate CellRanger input to be named, and the location of your Demuxlet .best output.

Examples:

```
library(Seurat)
library(DittoSeq)
#10X Data
Object <- Import10XDemux(CellRangerLocations = c("Location1/",
                                                 "Location2/",
                                                 "Location3/"),
                         Lane.names = c("Lane1","Lane2","Lane3"),
                         Demuxlet.best = "Location/Demuxlet.best",
                         verbose = TRUE)
#Seurat Objects (Any other type of data)
Object <- ImportSeuratDemux(Seurats = c(CD4, CD4.8, CD8),
                            Lane.names = c("CD4","CD4.8","CD8"),
                            Demuxlet.best = "Location/Demuxlet.best",
                            verbose = TRUE)
```

Both functions give quick summary of what was done at the end which will look like this:

```
[1] SUMMARY:
[1] 3 Seurat objects were combined with demuxlet data extracted into metadata.
[1] The average number of SNPs per cell was: 499.4
[1] Out of 49659 cells total, Demuxlet assigned:
[1]      40692 cells or 81.9% as singlets
[1]      8967 cells or 18.1% as doublets
[1]      0 cells as too ambiguous to call.
```

## Demuxlet Summary Plotting Functions

-	**`demux.calls.summary()`** - Makes a plot of how many calls were made per sample, separated by the separate lanes.  This is very useful for checking the potential accuracy of sample calls when only certain samples went into certain lanes/pools/sequencing runs/etc.  (Note: the default setting is to only show Singlet calls.  Use `singlets.only = FALSE` to include *one of the sample calls* for any doublets.

-	**`demux.SNP.summary()`** - Useful for checking if you have a lot of cells with very few SNPs. Creates a plot of the number of SNPs per cell that is grouped by individual lane by default.  This function is a simple wrapper for DBPlot() function with var="demux.N.SNP" and with a number of input defaults are adjusted (such as group.by and color.by = "Lane" so that the grouping is done according to 'Lane' metadata.

-	Both of these work with the same object=DEFAULT shortcut as all other DittoSeq functions, 

Examples:

```
library(Seurat)
library(DittoSeq)

#Method1
demux.SNP.summary(Seurat.object)
demux.calls.summary(Seurat.object)

#Method2
DEFAULT <- "Seurat.object"
demux.SNP.summary()
demux.calls.summary()
```

## Metadata created:

Metadata slot name | Description or the Demuxlet.best column name if directly carried over
--- | ---
Lane | guided by Lane.names import input, represents of separate droblet-generation lane, pool, sequencing lane, etc.
Sample | The sample call, from the BEST column
demux.doublet.call | whether the sample was a singlet (SNG), doublet (DBL), or ambiguious (AMB), from the BEST column
demux.RD.TOTL | RD.TOTL
demux.RD.PASS | RD.PASS
demux.RD.UNIQ | RD.UNIQ
demux.N.SNP | N.SNP
demux.PRB.DBL | PRB.DBL
demux.barcode.dup | (Only generated when TRUEs will exist) whether a cell's barcode refered to only 1 row of the .best file, but multiple distinct cells in the dataset (indicative of a technical issue in the bioinformatics pipeline)
