#' The RNAseq Class
#' @importFrom methods new slotNames
#' @importFrom stats median prcomp sd
#'
#' @description The RNAseq object stores RNAseq data in a structure that dittoSeq plotting functions know how to handle.
#' All that is needed to create an RNAseq object is a DESeqDataSet, the output from the DESeq() function.
#' In future, outputs of Limma Voom and EdgeR will work as well.
#' @slot counts a matrix. The raw genes x samples counts data. It is recommended, but not required, that one of these should be given when a new RNBAseq object is created.
#' @slot DEobject A DESeqDataSet. The output of having run DESeq() on your data.
#' @slot DEtype String, the type of object. (Will become more meaningful after support for Limma and EdgeR is added)
#' @slot data matrix, the normalized data, often the regularized log correction of the counts data generated by a call to DESeq's rlog function.
#' @slot meta.data a data.frame containing meta-information about each sample.
#' Autopopulated from the DESeqDataSet upon import, or added to manually afterward.
#' Can be sample names, conditions, timepoints, Nreads, etc.
#' @slot reductions a list of dimensional reductions that have been run.
#' Each should be a list containing 3 elements:
#' \itemize{
#' \item embeddings: a matrix containing the dimensional reduction values, with individual dimensions as the columns
#' \item key: How the components of this dimensionality reduction should be refered to in plot axes (ex: "PC" for pca)
#' \item raw.object: the original data from the initial calculation. Not used by any dittoSeq funcitons, but this is still a nice place to store the object to keep all the data together.
#' }
#' @slot samples a string vector, the names of the samples.
#' @slot misc A great place to store any other data associated with your bulk experiment that does not fit elsewhere in the object. Left \code{NULL} by default, and dittoSeq functions do not use or adjust this slot in any way.
#'
#' This slot can hold data of any type, including a list of multiple data objects.
#' @seealso
#' \code{\link{importDESeq2}} for automated import of DESeq objects into an \code{RNAseq} class object
#'
#' \code{\link{addMetaRNAseq}} for how to add metadata to an \code{RNAseq} object
#'
#' \code{\link{addPrcomp}} and \code{\link{addDimReduction}} for how to add dimensionality reductions to an \code{RNAseq} object for use in
#' \code{\link{dittoDimPlot}} visualizations.
#'
#' @author Daniel Bunis
#' @export

Class <- setClass(
    "RNAseq",
    representation(
        counts = "matrix",
        DEobject = "ANY",
        DEtype = "character",
        data = "matrix",
        meta.data = "data.frame",
        reductions = "list",
        samples = "character",
        misc = "ANY"
    ))

#' Creates an RNAseq object from a DESeq object.
#'
#' @description The first step of visualization of DESeq-analyzed bulk RNAseq data is running this function. Doing will extract meta.data information from the DESeq object, data using the rlog function, and if run_PCA=TRUE, will populate all other slots as well.
#' @param dds The output of running DESeq() on your data. = The DESeq2 object for your data. REQUIRED.
#' @param blind Logical which sets whether rlog estimation should be blinded to sample info. Run \code{\link[DESeq2]{rlog}} for more info about whether it should.
#' @param counts Matrix. The raw counts data matrix.  Not required but HIGHLY RECOMMENDED.
#' @param reductions For advanced users, a named list containing individual dimensionality reductions perviously run.
#' @return Outputs an \code{\linkS4class{RNAseq}} object.
#' @details This function takes in a \code{dds} DESeqDataSet (the output of running \code{DESeq()})
#' and generates an \code{\linkS4class{RNAseq}} object with
#' \code{counts}, \code{data}, \code{samples}, and \code{meta.data} automatically pulled and filled in from the \code{dds}.
#'
#' For the \code{data} slot, the user must first determine whether regularized log calculation should be blinded, and this decision must be supplied to the \code{blind} input.
#' For more information on that, see \code{\link[DESeq2]{rlog}}.
#' @seealso
#' \code{\link[DESeq2]{DESeq}} and \code{\link[DESeq2]{rlog}} for information about DESeq and regularized log calculation.
#'
#' \code{\link{addMetaRNAseq}} for how to add additional metadata to an \code{RNAseq} object
#'
#' \code{\link{addPrcomp}} and \code{\link{addDimReduction}} for how to add dimensionality reductions to an \code{\linkS4class{RNAseq}} object for use in
#' \code{\link{dittoDimPlot}} visualizations.
#' @examples
#'
#' # Generate mock RNAseq counts and a DESeq object from the mock data
#' # count tables from RNA-Seq data
#' counts.table <- matrix(rnbinom(n=1000, mu=100, size=1/0.5), ncol=10)
#' colnames(counts.table) <- paste0("Sample",1:10)
#' rownames(counts.table) <- paste0("Gene",1:100)
#' conditions <- factor(rep(1:2, each=5))
#'
#' # object construction
#' library(DESeq2)
#' dds <- DESeqDataSetFromMatrix(
#'     counts.table, DataFrame(conditions), ~ conditions)
#' dds <- DESeq(dds)
#'
#' # Import
#' myRNA <- importDESeq2(dds, blind = FALSE)
#'
#' # Add a dimensionality reduction to myRNA
#' #   NOTE: This is typically not done with all genes in a dataset.
#' #   The inclusion of this example code is not an endorsement of a particular
#' #   method of PCA. Consult yourself, a bioinformatician, or literature for
#' #   tips on proper techniques.
#' PCA <- prcomp(t(myRNA@data), center = TRUE, scale = TRUE)
#' myRNA <- addDimReduction(
#'     embeddings = PCA$x,
#'     object = myRNA,
#'     name = "pca",
#'     key = "PC",
#'     raw.object = PCA)
#'
#' ### Add a batch metadata
#' # Option 1, all in the funciton:
#' myRNA <- addMetaRNAseq(
#'     value = rep(1:2, each = 5),
#'     name = "batch",
#'     object = myRNA)
#' # Option 2, create variable beforehand, and just provide that to value.
#' #     The name of the variable will be used to name the metadata slot.
#' batch <- rep(1:2, each = 5)
#' myRNA <- addMetaRNAseq(batch, object = myRNA)
#'
#' # Visualize batch on a PCA plot
#' #   Note: For RNAseq objects, reduction.use defaults to "pca"
#' #   so just dittoDimPlot("Sample", myRNA, size = 3) would work the same!
#' dittoDimPlot("batch", myRNA, reduction.use = "pca", size = 3)
#'
#' @author Daniel Bunis
#' @export

importDESeq2 <- function(
    dds, blind, counts = NULL, reductions = NULL) {

    object <- new(
        "RNAseq", DEobject = dds, DEtype = "DESeq2")

    if (is.null(counts)) {
        object@counts <- DESeq2::counts(dds)
    } else {
        object@counts <- counts
    }

    if (!is.null(reductions)) {
        object@reductions <- reductions
    }

    object@samples <- colnames(object@counts)

    object@meta.data <- as.data.frame(SummarizedExperiment::colData(dds))
    if (is.null(object@meta.data$Sample)) {
        object@meta.data$Sample <- object@samples
    }
    if (is.null(object@meta.data$Nreads)) {
        object@meta.data$Nreads <- colSums(object@counts)
    }

    object@data <- SummarizedExperiment::assay(
        DESeq2::rlog(dds, blind = blind))

    object
}

#' Add a prcomp pca calculation to an RNAseq object
#'
#' @param prcomp a prcomp output which will be added to the \code{object}
#' @param object the \code{\linkS4class{RNAseq}} object. REQUIRED, unless \code{DEFAULT <- "object"} has been run.
#' @param name String name for the reduction slot. Example: "pca".
#' This will become the name of the slot, what should should be provided to the \code{reduction.use} input when making a \code{\link{dittoDimPlot}}.
#' @param key String, like "PC", which sets the default axes-label prefix when this reduction is used for making a \code{\link{dittoDimPlot}}
#' @return Outputs an \code{\linkS4class{RNAseq}} object with a new \code{@reductions$'name'} slot.
#' @seealso
#' \code{\link{addDimReduction}} for adding other types of dimensionality reductions
#'
#' \code{\link{addMetaRNAseq}} for how to add metadata to an \code{RNAseq} object
#'
#' \code{\link{importDESeq2}} for initial import of \code{RNAseq} data analyzed with DESeq2
#'
#' \code{\link{dittoDimPlot}} for visualizing how samples group within this PCA
#'
#' \code{\linkS4class{RNAseq}} for learning more about the \code{RNAseq} object type
#' @examples
#'
#' # Generate mock RNAseq counts and a DESeq object from the mock data
#' # count tables from RNA-Seq data
#' counts.table <- matrix(rnbinom(n=1000, mu=100, size=1/0.5), ncol=10)
#' colnames(counts.table) <- paste0("Sample",1:10)
#' rownames(counts.table) <- paste0("Gene",1:100)
#' conditions <- factor(rep(1:2, each=5))
#'
#' # object construction
#' library(DESeq2)
#' dds <- DESeqDataSetFromMatrix(
#'     counts.table, DataFrame(conditions), ~ conditions)
#' dds <- DESeq(dds)
#'
#' # Import
#' myRNA <- importDESeq2(dds, blind = FALSE)
#'
#' # Add PCA calculated with prcomp using addPrcomp
#' #   NOTE: This is typically not done with all genes in a dataset.
#' #   The inclusion of this example code is not an endorsement of a particular
#' #   method of PCA. Consult yourself, a bioinformatician, or literature for
#' #   tips on proper techniques.
#' myRNA <- addPrcomp(
#'     prcomp = prcomp(t(myRNA@data), center = TRUE, scale = TRUE),
#'     object = myRNA)
#'
#' ### Add a batch metadata
#' # Option 1, all in the funciton:
#' myRNA <- addMetaRNAseq(
#'     value = rep(1:2, each = 5),
#'     name = "batch",
#'     object = myRNA)
#' # Option 2, create variable beforehand, and just provide that to value.
#' #     The name of the variable will be used to name the metadata slot.
#' batch <- rep(1:2, each = 5)
#' myRNA <- addMetaRNAseq(batch, object = myRNA)
#'
#' # Visualize batch on a PCA plot
#' #   Note: For RNAseq objects, reduction.use defaults to "pca"
#' #   so just dittoDimPlot("Sample", myRNA, size = 3) would work the same!
#' dittoDimPlot("batch", myRNA, reduction.use = "pca", size = 3)
#'
#' @author Daniel Bunis
#' @export

addPrcomp <- function(prcomp, object = DEFAULT, name = "pca", key = "PC") {
    #Turn the "name" of the object into the object itself if name was given
    if (typeof(object)=="character") {
        object <- eval(expr = parse(text = object))
    }
    addDimReduction(prcomp$x,object,name,key,prcomp)
}

#' Add any dimensionality reduction space to an RNAseq object
#'
#' @param embeddings a numeric matrix containing the coordinates of all samples within the dimensionality reduction space.
#' @param object the \code{\linkS4class{RNAseq}} object. REQUIRED, unless \code{DEFAULT <- "object"} has been run.
#' @param name String name for the reduction slot. Example: "pca".
#' This will become the name of the slot, what should should be provided to the \code{reduction.use} input when making a \code{\link{dittoDimPlot}}.
#' @param key String, like "PC", which sets the default axes-label prefix when this reduction is used for making a \code{\link{dittoDimPlot}}.
#' If nothing is provided, a key will be automatically generated.
#' @param raw.object Optional, but recommended as this keeps the data all together: the output of the original calculation.
#' @return Outputs an \code{\linkS4class{RNAseq}} object with a new \code{@reductions$'name'} slot.
#' @seealso
#' \code{\link{addPrcomp}} for a prcomp specific PCA import wrapper function
#'
#' \code{\link{addMetaRNAseq}} for how to add metadata to an \code{RNAseq} object
#'
#' \code{\link{importDESeq2}} for initial import of \code{RNAseq} data analyzed with DESeq2
#'
#' \code{\link{dittoDimPlot}} for visualizing how samples group within this dimensionality reduction space after adding it to the object
#'
#' \code{\linkS4class{RNAseq}} for learning more about the \code{RNAseq} object type
#' @examples
#'
#' # Generate mock RNAseq counts and a DESeq object from the mock data
#' # count tables from RNA-Seq data
#' counts.table <- matrix(rnbinom(n=1000, mu=100, size=1/0.5), ncol=10)
#' colnames(counts.table) <- paste0("Sample",1:10)
#' rownames(counts.table) <- paste0("Gene",1:100)
#' conditions <- factor(rep(1:2, each=5))
#'
#' # object construction
#' library(DESeq2)
#' dds <- DESeqDataSetFromMatrix(
#'     counts.table, DataFrame(conditions), ~ conditions)
#' dds <- DESeq(dds)
#'
#' # Import
#' myRNA <- importDESeq2(dds, blind = FALSE)
#'
#' # Add a dimensionality reduction to myRNA
#' #   NOTE: This is typically not done with all genes in a dataset.
#' #   The inclusion of this example code is not an endorsement of a particular
#' #   method of PCA. Consult yourself, a bioinformatician, or literature for
#' #   tips on proper techniques.
#' PCA <- prcomp(t(myRNA@data), center = TRUE, scale = TRUE)
#' myRNA <- addDimReduction(
#'     embeddings = PCA$x,
#'     object = myRNA,
#'     name = "pca",
#'     key = "PC",
#'     raw.object = PCA)
#'
#' ### Add a batch metadata
#' # Option 1, all in the funciton:
#' myRNA <- addMetaRNAseq(
#'     value = rep(1:2, each = 5),
#'     name = "batch",
#'     object = myRNA)
#' # Option 2, create variable beforehand, and just provide that to value.
#' #     The name of the variable will be used to name the metadata slot.
#' batch <- rep(1:2, each = 5)
#' # Option 3, create variable beforehand, and provide just that using the
#' #     even further simplified setter function:
#' addMeta(myRNA) <- batch
#'
#' # Visualize batch on a PCA plot
#' #   Note: For RNAseq objects, reduction.use defaults to "pca"
#' #   so just dittoDimPlot("Sample", myRNA, size = 3) would work the same!
#' dittoDimPlot("batch", myRNA, reduction.use = "pca", size = 3)
#'
#' @author Daniel Bunis
#' @export

addDimReduction <- function(
    embeddings, object = DEFAULT, name, key = .gen_key(name),
    raw.object = NULL) {

    # Turn the "name" of the object into the object itself if name was given
    if (typeof(object)=="character") {
        object <- eval(expr = parse(text = object))
    }

    # Make the reduction
    new <- list(list(
        embeddings = embeddings,
        key = key,
        raw.object = raw.object))
    if (name %in% names(object@reductions)) {
        object@reductions[names(object@reductions) %in% name] <- new
    } else {
        object@reductions[length(object@reductions) + 1] <- new
        names(object@reductions)[length(object@reductions)] <- name
    }
    object
}

#' Add metadata slot to an RNAseq object
#'
#' @param value Vector of any type with length equal to the number of samples in the \code{object}
#' @param object the \code{\linkS4class{RNAseq}} object. REQUIRED, unless \code{DEFAULT <- "object"} has been run.
#' @param name String name for the metadata slot. Example: "age".
#' By default, the name of the object provided to value is used.
#'
#' This will become the name of the slot AND the string that should be provided to \code{var}-type inputs of dittoSeq plotting funcitons.
#' @return Outputs an \code{\linkS4class{RNAseq}} object with a new \code{@meta.data$'name'} slot.
#' @seealso
#' \code{\link{importDESeq2}} for initial import of \code{RNAseq} data analyzed with DESeq2
#'
#' \code{\link{addDimReduction}} and \code{\link{addPrcomp}} for adding dimensionality reductions
#'
#' \code{\linkS4class{RNAseq}} for learning more about the \code{RNAseq} object type
#' @examples
#'
#' # Generate mock RNAseq counts and a DESeq object from the mock data
#' # count tables from RNA-Seq data
#' counts.table <- matrix(rnbinom(n=1000, mu=100, size=1/0.5), ncol=10)
#' colnames(counts.table) <- paste0("Sample",1:10)
#' rownames(counts.table) <- paste0("Gene",1:100)
#' conditions <- factor(rep(1:2, each=5))
#'
#' # object construction
#' library(DESeq2)
#' dds <- DESeqDataSetFromMatrix(
#'     counts.table, DataFrame(conditions), ~ conditions)
#' dds <- DESeq(dds)
#'
#' # Import
#' myRNA <- importDESeq2(dds, blind = FALSE)
#'
#' # Add PCA with prcomp, but use generic functionto add to myRNA
#' #   NOTE: This is typically not done with all genes in a dataset.
#' #   The inclusion of this example code is not an endorsement of a particular
#' #   method of PCA. Consult yourself, a bioinformatician, or literature for
#' #   tips on proper techniques.
#' PCA <- prcomp(t(myRNA@data), center = TRUE, scale = TRUE)
#' myRNA <- addDimReduction(
#'     embeddings = PCA$x,
#'     object = myRNA,
#'     name = "pca",
#'     key = "PC",
#'     raw.object = PCA)
#'
#' ### Add a batch metadata
#' # Option 1, all in the funciton:
#' myRNA <- addMetaRNAseq(
#'     value = rep(1:2, each = 5),
#'     name = "batch",
#'     object = myRNA)
#' # Option 2, create variable beforehand, and just provide that to value.
#' #     The name of the variable will be used to name the metadata slot.
#' batch <- rep(1:2, each = 5)
#' myRNA <- addMetaRNAseq(batch, object = myRNA)
#'
#' # Visualize batch on a PCA plot
#' #   Note: For RNAseq objects, reduction.use defaults to "pca"
#' #   so just dittoDimPlot("Sample", myRNA, size = 3) would work the same!
#' dittoDimPlot("batch", myRNA, reduction.use = "pca", size = 3)
#'
#' @author Daniel Bunis
#' @export
addMetaRNAseq <- function(
    value, name = deparse(substitute(value)), object = NULL) {

    if (is.null(object) || is.character(object)) {
        stop("object must be given directly.")
    }
    if (name %in% names(object@meta.data)) {
        object@meta.data[names(object@meta.data) %in% name] <- value
    } else {
        object@meta.data[length(object@meta.data)+1] <- value
        names(object@meta.data)[length(object@meta.data)] <- name
    }
    object
}

#' @rdname addMetaRNAseq
#' @export
`addMeta<-` <- function(object, value)
{
    UseMethod('addMeta<-', object)
}

#' @rdname addMetaRNAseq
#' @export
`addMeta<-.RNAseq` <- function (object, value) {
    addMetaRNAseq(value, deparse(substitute(value)), object)
}


